enclave.go

```diff
@@ -309,16 +342,152 @@ func (e *Enclave) Start() error {
 		return fmt.Errorf("%s: %w", errPrefix, err)
 	}
 
+	if !e.cfg.isScalingEnabled() {
+		return nil
+	}
+
+	// Check if we are the leader.
+	if !e.weAreLeader(ctx) {
+		elog.Println("Obtaining worker's hostname.")
+		worker := getSyncURL(getHostnameOrDie(), e.cfg.ExtPrivPort)
+		err = asWorker(e.setupWorkerPostSync, e.attester).registerWith(leader, worker)
+		if err != nil {
+			elog.Fatalf("Error syncing with leader: %v", err)
+		}
+	}
+
 	return nil
 }
 
+func (e *Enclave) weAreLeader(ctx context.Context) (result bool) {
+	var (
+		err         error
+		ourNonce    nonce
+		weAreLeader = make(chan struct{}, 1)
+		areWeLeader = make(chan bool)
+		errChan     = make(chan error)
+		leader      = e.getLeader(pathLeader)
+	)
+	defer func() {
+		elog.Printf("We are leader: %v", result)
+	}()
+
+	ourNonce, err = newNonce()
+	if err != nil {
+		elog.Fatalf("Error creating new nonce: %v", err)
+	}
+
+	m := e.extPrivSrv.Handler.(*chi.Mux)
+	m.Get(pathLeader, getLeaderHandler(ourNonce, weAreLeader))
+	// Reset the handler as we no longer have a need for it.
+	defer m.Get(pathLeader,
+		func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(http.StatusGone)
+		},
+	)
+
+	timeout := time.NewTicker(10 * time.Second)
+	for {
+		go makeLeaderRequest(leader, ourNonce, areWeLeader, errChan)
+		select {
+		case <-errChan:
+			elog.Println("Not yet able to talk to leader designation endpoint.")
+			time.Sleep(time.Second)
+			continue
+		case result = <-areWeLeader:
+			return
+		case <-weAreLeader:
+			e.setupLeader(ctx)
+			result = true
+			return
+		case <-timeout.C:
+			elog.Fatal("Timed out talking to leader designation endpoint.")
+		}
+	}
+}
+
+// setupWorkerPostSync performs necessary post-key synchronization tasks like
+// installing the given enclave keys and starting the heartbeat loop.
+func (e *Enclave) setupWorkerPostSync(keys *enclaveKeys) error {
+	e.keys.set(keys)
+	cert, err := tls.X509KeyPair(keys.NitridingCert, keys.NitridingKey)
+	if err != nil {
+		return err
+	}
+	e.httpsCert.set(&cert)
+
+	// Start our heartbeat.
+	worker := getSyncURL(getHostnameOrDie(), e.cfg.ExtPrivPort)
+	go e.workerHeartbeat(context.Background(), worker)
+
+	return nil
+}
+
+// setupLeader performs necessary setup tasks like starting the worker event
+// loop and installing leader-specific HTTP handlers.
+func (e *Enclave) setupLeader(ctx context.Context) {
+	go e.workers.start(ctx)
+	// Make leader-specific endpoints available.
+	e.intSrv.Handler.(*chi.Mux).Put(pathState, putStateHandler(e))
```

Comment:
Ok, so all instances respond to `GET /enclave/state` on the internal port, but only the active leader responds to `PUT` on that endpoint.

We haven't really decided how the inner enclave application interacts with state updates. With the current code, it could alternating polling this endpoint with both `PUT` and `GET` for a certain amount of time and decide whether it's the leader based on that.

Is there a better way to handle the sequencing of this? How would it interact with the `/enclave/ready` notification from the inner app and the `--appcmd` switch to start the application. We have a similar issue when a worker receives new key material from the leader.